internal/coding/quotedprint.go

```diff
@@ -9,33 +9,87 @@ import (
 // QPCleaner scans quoted printable content for invalid characters and encodes them so that
 // Go's quoted-printable decoder does not abort with an error.
 type QPCleaner struct {
-	in *bufio.Reader
+	in       *bufio.Reader
+	overflow []byte
+	lineLen  int
 }
 
-// Assert QPCleaner implements io.Reader.
-var _ io.Reader = &QPCleaner{}
+// MaxQPLineLen is the maximum line length we allow before inserting `=\r\n`.  Prevents buffer
+// overflows in mime/quotedprintable.Reader.
+const MaxQPLineLen = 1024
+
+var (
+	_ io.Reader = &QPCleaner{} // Assert QPCleaner implements io.Reader.
+
+	escapedEquals = []byte("=3D") // QP encoded value of an equals sign.
+	lineBreak     = []byte("=\r\n")
+)
 
 // NewQPCleaner returns a QPCleaner for the specified reader.
 func NewQPCleaner(r io.Reader) *QPCleaner {
 	return &QPCleaner{
-		in: bufio.NewReader(r),
+		in:       bufio.NewReader(r),
+		overflow: nil,
+		lineLen:  0,
 	}
 }
 
 // Read method for io.Reader interface.
 func (qp *QPCleaner) Read(dest []byte) (n int, err error) {
-	// Ensure room to write a byte or =XX string.
-	destLen := len(dest) - 3
-	// Loop over bytes in qp.in ByteReader.
+	destLen := len(dest)
+
+	if len(qp.overflow) > 0 {
+		// Copy bytes that didn't fit into dest buffer during previous read.
+		n = copy(dest, qp.overflow)
+		qp.overflow = qp.overflow[n:]
+	}
+
+	// writeByte outputs a single byte, space for which will have already been ensured by the loop
+	// condition. Updates counters.
+	writeByte := func(in byte) {
+		dest[n] = in
+		n++
+		qp.lineLen++
+	}
+
+	// writeBytes outputs multiple bytes, storing overflow for next read. Updates counters.
+	writeBytes := func(in []byte) {
+		nc := copy(dest[n:], in)
+		if nc < len(in) {
+			// Stash unwritten bytes into overflow.
+			qp.overflow = append(qp.overflow, []byte(in[nc:])...)
+		}
+		n += nc
+		qp.lineLen += len(in)
+	}
+
+	// ensureLineLen ensures there is room to write `requested` bytes, preventing a line break being
+	// inserted in the middle of the escaped string.
+	ensureLineLen := func(requested int) {
+		if qp.lineLen+requested >= MaxQPLineLen {
+			writeBytes(lineBreak)
+			qp.lineLen = 0
+		}
+	}
+
+	// Loop over bytes in qp.in ByteReader while there is space in dest.
 	for n < destLen {
 		var b byte
 		b, err = qp.in.ReadByte()
 		if err != nil {
 			return n, err
 		}
+
+		if qp.lineLen >= MaxQPLineLen {
+			writeBytes(lineBreak)
+			qp.lineLen = 0
+		}
+
 		switch {
+		// Pass valid hex bytes through, otherwise escapes the equals symbol.
 		case b == '=':
-			// Pass valid hex bytes through.
+			ensureLineLen(2)
```

Comment:
I could be misreading, but shouldn't these calls be requesting `3` bytes?
```suggestion
			ensureLineLen(3)
```

Reply:
No, I see. Never mind this one.

Reply:
I had the same confusion when I wrote it.  :)  I've added clarification to ensureLinLen's comment.